### DISCUSSION TITLE
Pages implementing `getServerSideProps` are open to POST requests

### DISCUSSION THREAD

**Discussion by domeq** [vercel/next.js#43341] [Ideas]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.1.0: Sun Oct  9 20:15:09 PDT 2022; root:xnu-8792.41.9~2/RELEASE_ARM64_T6000
    Binaries:
      Node: 16.18.1
      npm: 8.19.2
      Yarn: N/A
      pnpm: N/A
    Relevant packages:
      next: 13.0.3-canary.2
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Describe the Bug

Any page that implements `getServerSideProps` suddenly stops responding with `405 Method Not Allowed` when requested with `POST` and starts handling the request same way as if it was `GET`.

This behaviour allows potential attackers to easily bypass any reverse-proxy like caching service e.g. Cloudflare, Cloudfront, Fastly.

Making a simple POST request to a heavy page will execute entire page's logic on every request causing increased billing and potential backend overload.

### Expected Behavior

As `getServerSideProps` is a function designed to be used for data fetching on server-side, it should have nothing to do with allowing `POST` requests reaching the backend.

`POST` requests for non-API requests should by default always return `405 Method Not Allowed`.

### Link to reproduction - Issues with a link to complete (but minimal) reproduction code will be addressed faster

https://github.com/domeq/nextjs-post-handling-reproduction

### To Reproduce

1. Run `npm ci && npm build && npm start`
2. Verify how a 'static' route behaves:
    - `curl -s -I http://localhost:3000` correctly returns `HTTP/1.1 200 OK` along with content 👍🏻 
    - `curl -s -I -X POST http://localhost:3000` correctly returns `HTTP/1.1 405 Method Not Allowed` 👍🏻 
3. Verify how a 'dynamic' route (route that implements `getServerSideProps`) behaves:
    - `curl -s -I http://localhost:3000/dynamic` correctly returns `HTTP/1.1 200 OK` along with content 👍🏻 
    - `curl -s -I -X POST http://localhost:3000/dynamic` **incorrectly returns `HTTP/1.1 200 OK` along with content** 👎🏻

**Comment by balazsorban44** [comment]
Hi, this is currently intentional and is relied upon by libraries like https://github.com/smeijer/next-runtime so we cannot change this currently. We could potentially introduce an experimental flag that opts out of this behavior if you set it, but will likely not change the default as it would be a breaking change.

**Reply by domeq** [reply]
Yeah, I understand that by now this has become a feature for some people for good and bad and changing the default behaviour would require a major release.

Being able to opt out of it via experimental flag would be more than good enough for a start, especially if it was clearly explained in the docs. 👍🏻
